### PR TITLE
DS-4433 Specify embeds

### PR DESCRIPTION
**Pull request by cwilper**
https://jira.lyrasis.org/browse/DS-4433

Adds support for specifying exact embeds to include in responses, e.g.

```
items/{uuid}?embed=bundles&embed=owningCollection

- or -

items/{uuid}?embed=bundles,owningCollection
```

It also adds support for composite projections, using a similar syntax for multiple values. See javadocs for details.

```
items/{uuid}?projection=p1&projection=p2

- or -

items/{uuid}?projection=p1,p2
```

### CODE REVIEW

**Comment by tdonohue** [comment]
@cwilper : this can now be rebased as #2625 was just merged (which this PR depends on)

**Comment by benbosman** [comment]
@cwilper I noticed the embeds are currently not taking the embed level into account

When using `/api/core/communities/<:uuid>?embed=subcommunities&embed=collections`, it embeds the collections and subcommunities of the sub communities as well, which is not relevant information.

When using `/api/core/items/<:uuid>?embed=logo&embed=owningCollection`, it embeds the collection logo (in this case there's no item logo, so it's a test request)

This would be easy to solve by using:
```
    public boolean allowEmbedding(HALResource<? extends RestAddressableModel> halResource, LinkRest linkRest) {
        if (halResource.getContent().getEmbedLevel() == 0)
            return embedRels.contains(linkRest.name());
        return false;
    }
```
The HALResource should use RestAddressableModel across the various Projection implementations, but would be easy to adjust

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/projection/EmbedRelsProjection.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+ */
+package org.dspace.app.rest.projection;
+
+import java.util.Set;
+
+import org.dspace.app.rest.model.LinkRest;
+import org.dspace.app.rest.model.hateoas.HALResource;
+
+/**
+ * Projection that allows a given set of rels to be embedded.
```
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/projection/EmbedRelsProjection.java]: Silly question, but what are "rels"?  Can we enhance the JavaDocs here a bit...it's unclear what constitutes a valid "rel". For example, how do we know what "rels" are valid to pass in?

That question is a bit "tongue in cheek", as I'm assuming that by "rels" you mean relationships, which are actually named `_links` in HAL format.  But, it'd be good to spell that out, or even consider renaming this `EmbedLinksProjection` (since any `link` is valid to pass to `embed`, I presume).  The rename here is optional, but I the the Javadocs do need improving here.

**Review by tdonohue — COMMENTED** [review]
@cwilper : Gave this a code review & test today.  Overall the code looks good (one minor comment inline).  But, I can reproduce some of the odd behaviors that @benbosman previously mentioned.  Here's just a few things that were unexpected (to me)...but I'm not sure whether we want this behavior or not:

* I can reproduce the odd behavior that @benbosman notes above.
* Similarly, a call to `/api/core/items/<:uuid>?embed=bundles,bitstreams` will embed all Bundles & Bitstreams for that Item.  Maybe that's nice if you want all that info, but it was unexpected to me.

Overall, I like this direction. It just seems like we need to add some better tests here to prove out the behavior is as expected, and that embeds are not being "inherited" from one level to another (that seems a bit odd to me, unless we find a real use case for that).

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/projection/EmbedRelsProjection.java (hunk 1 above)
Comment (cwilper) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/projection/EmbedRelsProjection.java]: "Rels" is common shorthand for "link relation types" as defined by RFC8288. The Spring HATEOAS folks call them that a lot (see the javadocs for "Link"), as do a few other places in DSpace code and in the wild (see the wikipedia page for "Link relation")...so I got used to doing the same.

It doesn't hurt to be explicit here, though, particularly in class-level javadocs.

> how do we know what "rels" are valid to pass in?

That info is expressed in a few places in code, but if you're consuming the REST api, your best bet for the official list should be the rest contract. The javadocs could stand to make reference to that, too.

**Comment by cwilper** [comment]
@tdonohue thanks for the review. As noted on the call earlier in the week, I spoke with Ben about this and he has some upcoming modifications for this PR, which will include additional tests.

**Comment by benbosman** [comment]
This PR has been updated to resolve my feedback above, and now corresponds with https://github.com/DSpace/Rest7Contract/pull/106

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/projection/EmbedRelsProjection.java (hunk 1 above)
Comment (tdonohue) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/projection/EmbedRelsProjection.java]: I'd rather we get a bit more specific in these JavaDocs, either via an example, a link to REST Contract or more explanation. I worry that assuming "rels" is a commonly understood term to all developers is a big assumption here.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/CommunityParentCommunityLinkRepository.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.rest.webmvc.ResourceNotFoundException;
+import org.springframework.security.access.prepost.PreAuthorize;
+import org.springframework.stereotype.Component;
+
+/**
+ * LinkRepository for the ParentCommunity object for a Community
+ */
+@Component(CommunityRest.CATEGORY + "." + CommunityRest.NAME + "." + CommunityRest.PARENT_COMMUNITY)
+public class CommunityParentCommunityLinkRepository extends AbstractDSpaceRestRepository
```
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/CommunityParentCommunityLinkRepository.java]: This class looks to be unrelated to this PR.  But, I see it's already on `master`.  Could you rebase this PR to make it easier to review the new changes?

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/projection/FullProjection.java [hunk 3]
```diff
@@ -18,14 +21,17 @@
 public class FullProjection extends AbstractProjection {
 
     public final static String NAME = "full";
+    private final int maxEmbed = DSpaceServicesFactory.getInstance().getConfigurationService()
+            .getIntProperty("projections.full.max", 2);
```
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/projection/FullProjection.java]: This new configuration does not appear to be added to a `*.cfg` file in this PR.  We should add it along with comments describing it.

It should probably be added to the existing `rest.cfg` file, and renamed `rest.projections.full.max`.  (NOTE: this same config is also used in SpecificLevelProjection)

File: dspace/config/spring/rest/projections.xml [hunk 4]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<beans xmlns="http://www.springframework.org/schema/beans" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="http://www.springframework.org/schema/beans
+           http://www.springframework.org/schema/beans/spring-beans-2.5.xsd">
+
+    <bean class="org.dspace.app.rest.projection.SpecificLevelProjection">
+        <property name="maxEmbed" value="5" />
```
Comment (tdonohue) [inline dspace/config/spring/rest/projections.xml]: Can we make this editable in the `rest.cfg` file, perhaps alongside `projections.full.max`?  Remember, any config can be referenced in Spring XML, so we could change this to be something like :
```
 <property name="maxEmbed" value="${rest.projection.specificLevel.maxEmbed}" />
```

**Review by tdonohue — COMMENTED** [review]
@benbosman : I haven't tested this yet, but I did a quick code review & added a few suggestions inline (all should be minor).  I also think this PR needs a rebase as I see a lot of changes in GitHub that are unrelated (and all seem to be from recently merged PRs) -- so, I suspect a rebase will clean those up and make this easier to review.

**Comment by tdonohue** [comment]
Thanks for the updates @Raf-atmire .  Could you or @benbosman  rebase this PR as well on `master`?  This seems to include unrelated commits which have already been merged into `master`, making this PR a bit harder to code review.

**Review by tdonohue — COMMENTED** [review]
I've now tested this, and it works as described in https://github.com/DSpace/Rest7Contract/pull/106  I haven't run into any issues.  So, this is looking good to me overall.

I would like to still wait to merge this until it has been rebased.  As noted above, GitHub seems to be displaying unrelated commits/changes in this PR, which makes a thorough code review difficult (it's hard to tell which parts of the code were changed).

**Comment by benbosman** [comment]
@tdonohue I've rebased the PR, the number of changes has indeed reduced quite a bit

**Review by tdonohue — APPROVED** [review]
👍 Re-reviewed and tested again today. This looks good now that it's been rebased!  Thanks @benbosman and @cwilper !

**Comment by tdonohue** [comment]
I've also retested this with https://github.com/DSpace/dspace-angular/pull/588 (Angular PR), and it all works well.  Merging this as it is small, works well and was flagged for one approval.
